collector.go

```diff
@@ -104,57 +114,55 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
 	log.Debugf("Connecting to %s, chroot=%s timeout=%s", c.zookeeper, config.Chroot, config.Timeout)
 	client, err := kazoo.NewKazoo(strings.Split(c.zookeeper, ","), &config)
 	if err != nil {
-		msg := fmt.Sprintf("Connection error: %s", err)
-		log.Error(msg)
-		ch <- prometheus.NewInvalidMetric(prometheus.NewDesc("zookeeper_error", msg, nil, nil), err)
+		log.Errorf("Connection error: %s", err)
+		// If we can't connect, there's nothing left to do but return the error to the client
+		ch <- prometheus.MustNewConstMetric(c.metrics.zookeeperScrapeError, prometheus.GaugeValue, 1)
 		return
 	}
-
 	defer client.Close()
 
 	c.clusterMetrics(ch, client)
 
+	wg := sync.WaitGroup{}
 	topics, err := client.Topics()
 	if err != nil {
-		msg := fmt.Sprintf("Error collecting list of topics: %s", err)
-		log.Error(msg)
-		ch <- prometheus.NewInvalidMetric(prometheus.NewDesc("zookeeper_topic_list_error", msg, nil, nil), err)
-		return
+		log.Errorf("Error collecting list of topics: %s", err)
+		atomic.StoreInt32(&c.zkErr, 1)
+	} else {
+		wg.Add(len(topics))
+		for _, topic := range topics {
+			go func(cl *collector, cz chan<- prometheus.Metric, t *kazoo.Topic) {
+				defer wg.Done()
+				if len(cl.topics) > 0 && !stringInSlice(t.Name, cl.topics) {
+					// skip topic if it's not on the list of topic to collect
+					log.Debugf("Skipping topic '%s', not in list: %s [%d]", t.Name, cl.topics, len(cl.topics))
+					return
+				}
+				c.topicMetrics(ch, t)
+			}(c, ch, topic)
+		}
 	}
 
 	consumers, err := client.Consumergroups()
 	if err != nil {
-		msg := fmt.Sprintf("Error collecting list of consumers: %s", err)
-		log.Error(msg)
-		ch <- prometheus.NewInvalidMetric(prometheus.NewDesc("zookeeper_consumer_list_error", msg, nil, nil), err)
-		return
+		log.Errorf("Error collecting list of consumers: %s", err)
+		atomic.StoreInt32(&c.zkErr, 1)
+	} else {
+		wg.Add(len(consumers))
+		for _, consumer := range consumers {
+			go func(cl *collector, cz chan<- prometheus.Metric, cg *kazoo.Consumergroup) {
+				defer wg.Done()
+				if len(cl.consumers) > 0 && !stringInSlice(cg.Name, cl.consumers) {
+					// skip consumer if it's not on the list of consumers to collect
+					log.Debugf("Skipping consumer '%s', not in list: %s [%d]", cg.Name, cl.consumers, len(cl.consumers))
+					return
+				}
+				c.consumerMetrics(ch, cg)
+			}(c, ch, consumer)
```

Comment:
Since it's the idiom in the existing code, and network latency is the main determining factor for scrape time, I'm going to defer that optimization until later.

Reply:
Done.